src/site/content/en/blog/webassembly-threads/index.md

```diff
@@ -0,0 +1,472 @@
+generic `wasm32-unknown-unknown` target for generic WebAssembly output.
+
+If Wasm is intended to be used in a Web environment, any interaction with JavaScript APIs is left to
+external libraries and tooling like [wasm-bindgen](https://rustwasm.github.io/docs/wasm-bindgen/)
+and [wasm-pack](https://rustwasm.github.io/docs/wasm-pack/). Unfortunately, this means that the
+standard library is not aware of Web Workers and standard APIs like
+[`std::thread`](https://doc.rust-lang.org/std/thread/) won't work when compiled to WebAssembly.
+
+Luckily, the majority of the ecosystem depends on higher-level libraries to take care of
+multithreading. At that level it's much easier to abstract away all the platform differences.
+
+In particular, [Rayon](https://crates.io/crates/rayon) is the most popular choice for
+data-parallelism in Rust. It allows to take method chains on regular iterators and, usually with a
+single line change, convert them in a way where they'd run in parallel on all available threads
+instead of sequentially. For example:
+
+```rust/3/2
+pub fn sum_of_squares(numbers: &[i32]) -> i32 {
+  numbers
+  .iter()
+  .par_iter()
+  .map(x => x * x)
+  .sum()
+}
+```
+
+This unassuming code will split up the input data, calculate `x * x` and partial sums in parallel
+threads, and in the end add up those partial results together.
+
+In order to accomodate for platforms without working `std::thread`, Rayon provides hooks that allow
+to define custom logic for spawning and exiting threads. We've tapped into those hooks and built an
+adapter that you can use with wasm-bindgen to spawn WebAssembly threads as Web Workers:
+[wasm-bindgen-rayon](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon).
+
+To use it, you need to re-export an `initThreadPool` function from the wasm-bindgen-rayon library in
+your main code and invoke it right after initialization.
+
+```js
+import init, { initThreadPool /* ... */ } from './pkg/index.js';
+
+// Regular wasm-bindgen initialization.
+await init();
+
+// Thread pool initialization with the given number of threads
+// (pass `navigator.hardwareConcurrency` if you want to use all cores).
+await initThreadPool(navigator.hardwareConcurrency);
+
+// ...now you can invoke any exported functions as you normally would
+```
+
+It will create a pool of Workers and reuse them throughout the lifetime of the program. This
+initialization is similar to the `-s PTHREAD_POOL_SIZE=...` option in Emscripten explained earlier,
+and needs to happen before the main code to avoid the same deadlocks.
+
+There are few other configurations to set up before you can use multithreading, so make sure to
+check the [detailed documentation](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#usage) for
+complete setup steps.
+
+{% Aside 'caution' %} The same
+[caveats](https://github.com/GoogleChromeLabs/wasm-bindgen-rayon#caveats) about blocking the main
+thread apply here too. {% endAside %}
```

Comment:
This is another caution that leads into the text, the same comment as above applies.